demo

```diff
@@ -55,16 +68,118 @@ if [[ "$?" -eq 1 ]]; then
     docker-compose pull ${compose_args}
 fi
 
-if [[ "$1" == "up" ]]; then
-    echo "Starting Mender demo environment.."
-fi
-
 # Pass this value on to the GUI container as an env variable
 export INTEGRATION_VERSION=$(git describe --tags --abbrev=0)
 
-exec docker-compose \
-     -f docker-compose.yml \
-     -f docker-compose.storage.minio.yml \
-     -f docker-compose.demo.yml \
-     $CLIENT \
-     "$@"
+if [[ "$1" != "up" ]]; then
+    exec docker-compose \
+         -f docker-compose.yml \
+         -f docker-compose.storage.minio.yml \
+         -f docker-compose.demo.yml \
+         $client \
+         "$@"
+fi
+
+# ------------------------------------------------------------------------------
+#
+#       The following code will only be run in the case ./demo up [args]
+#
+# ------------------------------------------------------------------------------
+
+echo "Starting the Mender demo environment..."
+
+docker-compose \
+    -f docker-compose.yml \
+    -f docker-compose.storage.minio.yml \
+    -f docker-compose.demo.yml \
+    $client \
+    "$@" -d
+
+# DC_PID=$!
+
+USER='myusername@example.com'; PASSWORD='mysecretpassword'
+MENDER_SERVER_USER="${USER}:${PASSWORD}"
+MENDER_SERVER_URI='https://localhost'
+
+# Block until the useradm service returns an HTTP response.
+echo "Waiting for the useradm service to start up"
+curl --silent -k -X POST -u $MENDER_SERVER_USER\
+     --fail\
+     --retry-connrefused\
+     --retry 60\
+     --retry-delay 10\
+     --retry-max-time 120\
+     $MENDER_SERVER_URI/api/management/v1/useradm/auth/login > /dev/null
+
+retval=$?
+case $retval in
+    22) ;; # Server 400 error, ie, server is up - continue.
+    0)  ;; # User exists - continue.
+    *)
+       echo "Failed to access the mender-useradm service within the allotted two minutes."
+       echo "Please check the docker-compose setup."
+       echo "exit code" $retval
+       exit $retval
+       ;;
+esac
+
+echo "Creating a new user..."
+
+# Create a new user.
+docker exec integration_mender-useradm_1 /usr/bin/useradm create-user --username=${USER} --password=${PASSWORD}
```

Comment:
@estenberg: This is something we didn't think about: If the demo artifact is to be uploaded, then a user is needed first, which means we can no longer wait for it to be done manually. Is this OK? I suppose we can just remove this step. One step less in the instructions is actually better, right?

Reply:
Indeed, makes sense.

I have a couple of recommendations, though:
* use mender-demo@example.com as user email (for cosmetics)
* generate random password of 12 characters & output it to the user (for security)
* ouput a link to where user can log in to change user name and password

I am not sure how feasible this is, and am aware that this is "just" a demo server. However, default credentials is probably the most common way to get compromised and I don't think we should fall in this pitfall.

Reply:
Smart! However, should the password remain the same over multiple sessions, or a new one generated at each `up` command invocation?

Reply:
I think if the user already exists, we can assume that the artifact is already uploaded and bail out early. This also saves us from complicated and almost certainly insecure methods of caching the generated password.